Change `operator==` overload that constrains to `is_scalar_v` to constrain to compatible types.

fsandhei opened the pull request:
https://github.com/nlohmann/json/issues/4165 shows an issue with the overload resolution when attempting to compare a `std::string` with `json` from left to right. This can be fixed by changing the constraint to instead of focusing on scalar types, we constrain the comparison operator to "compatible types", in par with the catch-call constructor.

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [ ]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header files `single_include/nlohmann/json.hpp` and `single_include/nlohmann/json_fwd.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

## Please don't

- The C++11 support varies between different **compilers** and versions. Please note the [list of supported compilers](https://github.com/nlohmann/json/blob/master/README.md#supported-compilers). Some compilers like GCC 4.7 (and earlier), Clang 3.3 (and earlier), or Microsoft Visual Studio 13.0 and earlier are known not to work due to missing or incomplete C++11 support. Please refrain from proposing changes that work around these compiler's limitations with `#ifdef`s or other means.
- Specifically, I am aware of compilation problems with **Microsoft Visual Studio** (there even is an [issue label](https://github.com/nlohmann/json/issues?utf8=✓&q=label%3A%22visual+studio%22+) for this kind of bug). I understand that even in 2016, complete C++11 support isn't there yet. But please also understand that I do not want to drop features or uglify the code just to make Microsoft's sub-standard compiler happy. The past has shown that there are ways to express the functionality such that the code compiles with the most recent MSVC - unfortunately, this is not the main objective of the project.
- Please refrain from proposing changes that would **break [JSON](https://json.org) conformance**. If you propose a conformant extension of JSON to be supported by the library, please motivate this extension.
- Please do not open pull requests that address **multiple issues**.

fsandhei committed:
Replace scalar type comparison restriction with one extended for compatible types.

fsandhei committed:
Create a concept for compatible types and use it instead of SFINAE.

fsandhei committed:
Run amalgamate script

fsandhei committed:
Add check for comparison from left to right.

nlohmann commented:
Can you please try to amalgamate with an older version of astyle, like astyle 3.2? There was a breaking change in astyle which makes the output differ unfortunately.

fsandhei commented:
> Can you please try to amalgamate with an older version of astyle, like astyle 3.2? There was a breaking change in astyle which makes the output differ unfortunately.

Sure, I'll do it some time later today after work.

fsandhei committed:
Try amalgamate again.

There seems to be some issues or inconsistencies with the astyle
version 3.2.1 from the AUR versus 3.2.1-build from apt:
   * --squeeze-lines is not defined in the one from Arch, so I had to
     comment that one out in order to make the amalgamate step run.
   * A bunch of files are affected by the amalgamate, which does not
     seem right.

fsandhei commented:
First, sorry, @nlohmann, for the trouble I'm causing here.

So, I tried amalgamating again. There seems to be some issues or inconsistencies with the astyle version 3.2.1 from the AUR versus 3.2.1-build from apt:
   * --squeeze-lines is not defined in the one from Arch, so I had to
     comment that one out in order to make the amalgamate step run.
  * As far as I see, `squeeze-line` was introduced in 3.3.x.
   * A bunch of files are affected by the amalgamate, which does not
     seem right.

Would you want me to perhaps unstage all the other unrelated files that got affected by this?

fsandhei committed:
Revert "Try amalgamate again."

This reverts commit ca407117ed6b346005a4787252cdbb6a94161b93.

nlohmann commented:
This is what I am using locally (patched Makefile):

```Makefile
# call the Artistic Style pretty printer on all source files
pretty:
	/Users/niels/Downloads/astyle/build/astyle \
	    --style=allman \
	    --indent=spaces=4 \
	    --indent-modifiers \
	    --indent-switches \
	    --indent-preproc-block \
	    --indent-preproc-define \
	    --indent-col1-comments \
	    --pad-oper \
	    --pad-header \
	    --align-pointer=type \
	    --align-reference=type \
	    --add-braces \
	    --convert-tabs \
	    --close-templates \
	    --lineend=linux \
	    --preserve-date \
	    --suffix=none \
	    --formatted \
	   $(SRCS) $(TESTS_SRCS) $(AMALGAMATED_FILE) $(AMALGAMATED_FWD_FILE) docs/examples/*.cpp
```

with Astyle 3.1.

fsandhei committed:
amalgamate with astyle==3.1.2

fsandhei commented:
Thank you for so quick response. The amalgamation seems happy now.

nlohmann commented:
> Thank you for so quick response. The amalgamation seems happy now.

Thanks for your effort!

fsandhei committed:
Don't use concept. Causes compilation error on GCC 10

Opting out of using concept and rather defining a structure while still
being able to not use SFINAE and instead `requires` expression.

Compiling with C++23 on GCC 13.2.1 fails on overload resolution on
calling operator== for `std::nullptr_t`. The solution for this was to
simply add another overload for nullptr, on par with the equivalent
`nullptr_t` constructor.

fsandhei commented:
I see that the GCC 10 CI run failed. It seems to be not able to parse the concept, which is odd as it [should be supported in GCC 10](https://en.cppreference.com/w/cpp/compiler_support/20).

For the sake of getting this in and have the least amount of hassle I don't use `concept`.

However I was pondering on the idea of having a `concept` which could replace these type traits definitions for recognizing "compatible types". I tinkered with this simple concept below. It worked perfectly fine. I don't know if it is missing any details, though. Would appreciate hearing your opinion on this, as I'm far from an expert :)

```cpp
#ifdef JSON_HAS_CPP_20
   #include <concepts>
    template <typename T, typename BasicJsonType, typename U = uncvref_t<T>>
    concept CompatibleType = requires(const U& type, BasicJsonType& j) {
        { to_json(j, type) } -> std::convertible_to<void>;
    };
#endif
```

The use of this would be quite simple, as we'd then can constrict constructors / operators with this concept:
```
template <CompatibleType<basic_json_t> T>
operator==(T rhs) const noexcept
{
   ...
}
```

EDIT: I realize I goofed myself. It is supported but `JSON_HAS_CPP_20` is not defined because `g++10` requires `-std=c++2a`. I'm able to get past this by also checking for `__cpp_concepts` to be defined at where we define the concept.

fsandhei committed:
Use SFINAE instead of 'requires' clause because of cppcheck.

fsandhei committed:
Use bool instead of auto

fsandhei commented:
I'm not really sure what is Codacy is actually complaining about. The struct member is being used in the type trait. It might be a bug in Codacy. Do you have any suggestions @nlohmann?

nlohmann commented:
> I'm not really sure what is Codacy is actually complaining about. The struct member is being used in the type trait. It might be a bug in Codacy. Do you have any suggestions @nlohmann?

Indeed looks like a false positive. I set it to "ignore".

fsandhei committed:
Amalgamate

fsandhei committed:
Fix compile errors on other versions of clang

fsandhei committed:
Operator for non equality

fsandhei committed:
amalgamate
